modules/dag-l1/src/main/scala/org/tessellation/dag/l1/domain/block/BlockProcessor.scala

```diff
@@ -0,0 +1,104 @@
+import cats.syntax.functor._
+import cats.syntax.semigroup._
+import cats.syntax.traverse._
+
+import scala.util.control.NoStackTrace
+
+import org.tessellation.dag.block.BlockValidator
+import org.tessellation.dag.block.BlockValidator.BlockValidationError
+import org.tessellation.dag.domain.block.DAGBlock
+import org.tessellation.dag.l1.domain.address.storage.AddressStorage
+import org.tessellation.dag.l1.domain.block.BlockProcessor.{BalanceNegative, BlockFailedValidation}
+import org.tessellation.dag.l1.domain.transaction.TransactionService
+import org.tessellation.kryo.KryoSerializer
+import org.tessellation.schema.address.Address
+import org.tessellation.schema.balance.Balance
+import org.tessellation.schema.transaction.Transaction
+import org.tessellation.security.signature.Signed
+import org.tessellation.security.{Hashed, SecurityProvider}
+
+import eu.timepit.refined.auto._
+import eu.timepit.refined.numeric.NonNegative
+import eu.timepit.refined.refineV
+import io.estatico.newtype.ops.toCoercibleIdOps
+
+class BlockProcessor[F[_]: Async: KryoSerializer: SecurityProvider](
+  addressStorage: AddressStorage[F],
+  blockService: BlockService[F],
+  blockValidator: BlockValidator[F],
+  transactionService: TransactionService[F]
+) {
+
+  def accept(signedBlock: Signed[DAGBlock]): F[Unit] =
+    for {
+      _ <- blockValidator.validate(signedBlock).map {
+        case Valid(_)         => Applicative[F].unit
+        case Invalid(reasons) => BlockFailedValidation(reasons).raiseError[F, Unit]
+      }
+
+      hashedBlock <- signedBlock.hashWithSignatureCheck.flatMap(_.liftTo[F])
+      hashedTransactions <- signedBlock.value.transactions.toList
+        .traverse(_.hashWithSignatureCheck.flatMap(_.liftTo[F]))
+        .map(_.toSet)
+
+      _ <- acceptTransactions(hashedTransactions)
+      _ <- blockService.accept(hashedBlock)
+      _ <- blockService.handleTipsUpdate(hashedBlock)
+    } yield ()
+
+  private def prepareBalances(transactions: Set[Transaction]): F[Map[Address, Balance]] = {
+    val sources = transactions.groupBy(_.source)
+    val destinations = transactions.groupBy(_.destination)
+    val all = sources.combine(destinations)
+    val addressBalanceChanges = all.map {
+      case (address, txs) =>
+        val change = txs.foldLeft(BigInt(0L))(
+          (change, tx) =>
+            tx match {
+              case Transaction(`address`, _, amount, fee, _, _) => change - amount.coerce - fee.coerce
+              case Transaction(_, `address`, amount, _, _, _)   => change + amount.coerce
+              case _                                            => change
+            }
+        )
+
+        address -> change
+    }
+
+    addressBalanceChanges.toList.traverse {
+      case (address, balanceChange) =>
+        for {
+          current <- addressStorage.getBalance(address)
+          newBalance <- refineV[NonNegative]
+            .apply(current.coerce + balanceChange)
+            .bimap(BalanceNegative, Balance(_))
```

Comment:
How balance can be negative if transactions are validated upfront (or should be) and applied in chain? A transaction that spends more is simply invalid and should be already rejected

Reply:
`refineV.apply` returns an `Either[String, Refined]` so I need to handle that. If there is no issue in validations and flow this error should never be thrown. If it throws the reason would be that the Balance got negative hence the name.

Reply:
I am wondering if we can omit that validation if we set both `current` and `balanceChange` as `NonNegative`

Reply:
Balance change is a `BigInt` and it may be negative. If you send dag to someone, the balance change will be negative.